bt_ddos_shield/manifest_manager.py

```diff
@@ -1,36 +1,254 @@
+    """
+    pass
+
+
+class ManifestNotFoundException(ManifestManagerException):
+    """
+    Exception thrown when manifest file is not found under given address.
+    """
+    pass
+
+
+@dataclass
+class Manifest:
+    """
+    Class representing manifest file containing encrypted addresses for validators.
+    """
+
+    encrypted_address_mapping: dict[Hotkey, bytes]
+    """ Mapping with addresses for validators (validator HotKey -> encrypted address) """
+    md5_hash: str
+    """ MD5 hash of the manifest data """
+
+
+class AbstractManifestSerializer(ABC):
+    """
+    Class used to serialize and deserialize manifest file.
+    """
+
+    @abstractmethod
+    def serialize(self, manifest: Manifest) -> bytes:
+        """
+        Serialize manifest. Output format depends on the implementation.
+        """
+        pass
+
+    @abstractmethod
+    def deserialize(self, serialized_data: bytes) -> Manifest:
+        """
+        Deserialize manifest. Throws ManifestDeserializationException if data format is not recognized.
+        """
+        pass
+
+
+class JsonManifestSerializer(AbstractManifestSerializer):
+    """
+    Manifest serializer implementation which serialize manifest to Json.
+    """
+
+    encoding: str
+
+    def __init__(self, encoding: str = "utf-8"):
+        """
+        Args:
+            encoding: Encoding used for transforming Json string to bytes.
+        """
+        self.encoding = encoding
+
+    def serialize(self, manifest: Manifest) -> bytes:
+        data: dict = {
+            "encrypted_address_mapping": manifest.encrypted_address_mapping,
+            "md5_hash": manifest.md5_hash
+        }
+        json_str: str = json.dumps(data, default=self._custom_encoder)
+        return json_str.encode(encoding=self.encoding)
+
+    def deserialize(self, serialized_data: bytes) -> Manifest:
+        try:
+            json_str: str = serialized_data.decode(encoding=self.encoding)
+            data = json.loads(json_str, object_hook=self._custom_decoder)
+            return Manifest(data["encrypted_address_mapping"], data["md5_hash"])
+        except Exception as e:
+            raise ManifestDeserializationException(f"Failed to deserialize manifest data: {e}") from e
+
+    @staticmethod
+    def _custom_encoder(obj: Any) -> Any:
+        if isinstance(obj, Hotkey):
+            return str(obj)
+
+        if isinstance(obj, bytes):
+            return base64.b64encode(obj).decode()
+
+    @staticmethod
+    def _custom_decoder(json_mapping: dict[str, Any]) -> Any:
+        if "encrypted_address_mapping" in json_mapping:
+            decoded_mapping: dict[Hotkey, bytes] = {}
+            for hotkey, encoded_address in json_mapping["encrypted_address_mapping"].items():
+                decoded_mapping[Hotkey(hotkey)] = base64.b64decode(encoded_address.encode())
+            json_mapping["encrypted_address_mapping"] = decoded_mapping
+        return json_mapping
 
 
 class AbstractManifestManager(ABC):
```

Comment:
Hmm, I wanted to have serializer objects to allow different serializations (if needed) for Manifest and Address classes. If serialization would be inside these classes, then I need to add abstract serialize/deserialize methods there and implement it in subclasses (or eventually have default serialize/deserialize in these base classes). Anyway this needs subclassing of simple Address/Manifest dataclass and I think pattern with serializer class looks nicer, but maybe I'm wrong. Or maybe you just want to remove abstract serialization?

Another thing is, if I will remove these serializer objects, then I should pass Address/Manifest class object in constructors of AbstractBlockchainManager/AbstractManifestManager to allow calling deserialize method from proper class? Or pass there only deserialize method? For serialization it is easy, because we have instance of class to call serialize on it, but when calling deserialize, we need to know which method to use.

Reply:
I thought about asking Github Copilot about it and got such response:

Q: Is it better to have serializer class or have deserialize and serialize methods inside serialized object?

A: It is generally better to have a separate serializer class rather than having deserialize and serialize methods inside the serialized object. This approach follows the Single Responsibility Principle, making the code more modular, easier to maintain, and test. It also allows for different serialization strategies without modifying the serialized object.

Here is a brief comparison:

Separate Serializer Class

Pros:  
Adheres to the Single Responsibility Principle.
Easier to maintain and test.
Allows for different serialization strategies.
Keeps the serialized object focused on its core responsibilities.
Cons:  
Requires additional classes and interfaces.

Methods Inside Serialized Object

Pros:  
Simpler structure with fewer classes.
Serialization logic is directly associated with the object.
Cons:  
Violates the Single Responsibility Principle.
Harder to maintain and test.
Difficult to change serialization strategy without modifying the object.

In summary, using a separate serializer class is a more flexible and maintainable approach.